shut this error up

```
Run yarn test
yarn run v1.22.18
$ yarn run hardhat test
$ /home/runner/work/podcast-nft/podcast-nft/node_modules/.bin/hardhat test
Error HH8: There's one or more errors in your config file:

  * Invalid value undefined for HardhatConfig.networks.rinkeby.url - Expected a value of type string.
  * Invalid account: #0 for network: rinkeby - Expected string, received undefined
  * Invalid value undefined for HardhatConfig.networks.mainnet.url - Expected a value of type string.
  * Invalid account: #0 for network: mainnet - Expected string, received undefined
  
To learn more about Hardhat's configuration, please go to https://hardhat.org/config/

For more info go to https://hardhat.org/HH[8](https://github.com/henkaku-center/podcast-nft/runs/6139200876?check_suite_focus=true#step:8:8) or run Hardhat with --show-stack-traces
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Error: Process completed with exit code 1.
```

--- a/.github/workflows/pull_request.yaml
+++ b/.github/workflows/pull_request.yaml
@@ -17,6 +17,12 @@ jobs:
         with:
           node-version: '16'
           cache: yarn
+      - name: Setup dotenv
+        # Test account's fixed private key in hardhat local node
+        run: |
+          echo 'PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80' > .env
+          echo 'STAGING_ALCHEMY_KEY=' >> .env
+          echo 'PROD_ALCHEMY_KEY=' >> .env
       - name: Install packages
         run: yarn
       - name: Lint

--- a/contracts/PodCastNFT.sol
+++ b/contracts/PodCastNFT.sol
@@ -2,13 +2,13 @@
 pragma solidity ^0.8.0;
 
 import "hardhat/console.sol";
-import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
 import "@openzeppelin/contracts/access/Ownable.sol";
 import "@openzeppelin/contracts/utils/Counters.sol";
 
 import {IERC20} from "./interface/IERC20.sol";
 
-contract PodCastNFT is ERC721, ERC721Enumerable, Ownable {
+contract PodCastNFT is ERC721, Ownable {
     using Counters for Counters.Counter;
     Counters.Counter private _tokenIds;
     IERC20 public henkakuToken;
@@ -46,23 +46,6 @@ contract PodCastNFT is ERC721, ERC721Enumerable, Ownable {
     }
 
     // override
-
-    function _beforeTokenTransfer(
-        address from,
-        address to,
-        uint256 tokenId
-    ) internal override(ERC721, ERC721Enumerable) {
-        super._beforeTokenTransfer(from, to, tokenId);
-    }
-
-    function supportsInterface(bytes4 interfaceId)
-        public
-        view
-        override(ERC721, ERC721Enumerable)
-        returns (bool)
-    {
-        return super.supportsInterface(interfaceId);
-    }
 
     function tokenURI(uint256 tokenId)
         public
@@ -298,5 +281,9 @@ contract PodCastNFT is ERC721, ERC721Enumerable, Ownable {
         );
         require(success, "Transaction faild");
         userAttribute[msg.sender].claimableToken = 0;
+    }
+
+    function totalSupply() public view returns (uint256) {
+        return _tokenIds.current();
     }
 }

--- a/contracts/PodCastNFT.sol
+++ b/contracts/PodCastNFT.sol
@@ -3,15 +3,13 @@ pragma solidity ^0.8.0;
 
 import "hardhat/console.sol";
 import "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
 import "@openzeppelin/contracts/access/Ownable.sol";
 import "@openzeppelin/contracts/utils/Counters.sol";
-import "./ConsensusAdminable.sol";
-import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
 
 import {IERC20} from "./interface/IERC20.sol";
-import {Base64} from "./libraries/Base64.sol";
 
-contract PodCastNFT is ERC721URIStorage, ERC721Enumerable, Ownable, ConsensusAdminable {
+contract PodCastNFT is ERC721URIStorage, ERC721Enumerable, Ownable {
     using Counters for Counters.Counter;
     Counters.Counter private _tokenIds;
     IERC20 public henkakuToken;
@@ -42,6 +40,39 @@ contract PodCastNFT is ERC721URIStorage, ERC721Enumerable, Ownable, ConsensusAdm
         henkakuToken = IERC20(_erc20);
         setPrice(1000e18);
         setFundAddress(_fundAddress);
+    }
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal override(ERC721, ERC721Enumerable) {
+        super._beforeTokenTransfer(from, to, tokenId);
+    }
+
+    function _burn(uint256 tokenId)
+        internal
+        override(ERC721, ERC721URIStorage)
+    {
+        super._burn(tokenId);
+    }
+
+    function tokenURI(uint256 tokenId)
+        public
+        view
+        override(ERC721, ERC721URIStorage)
+        returns (string memory)
+    {
+        return super.tokenURI(tokenId);
+    }
+
+    function supportsInterface(bytes4 interfaceId)
+        public
+        view
+        override(ERC721, ERC721Enumerable)
+        returns (bool)
+    {
+        return super.supportsInterface(interfaceId);
     }
 
     function setPrice(uint256 _price) public onlyOwner {
@@ -119,19 +150,6 @@ contract PodCastNFT is ERC721URIStorage, ERC721Enumerable, Ownable, ConsensusAdm
             _point
         );
         _setTokenURI(tokenId, finalTokenUri);
-    }
-
-    function supportsInterface(bytes4 interfaceId)
-        public
-        view
-        virtual
-        override(ERC721)
-        returns (bool)
-    {
-        return
-            interfaceId == type(IERC721).interfaceId ||
-            interfaceId == type(IERC721Metadata).interfaceId ||
-            super.supportsInterface(interfaceId);
     }
 
     function getTokenURI(

--- a/contracts/PodCastNFT.sol
+++ b/contracts/PodCastNFT.sol
@@ -2,17 +2,17 @@
 pragma solidity ^0.8.0;
 
 import "hardhat/console.sol";
-import "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
 import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
 import "@openzeppelin/contracts/access/Ownable.sol";
 import "@openzeppelin/contracts/utils/Counters.sol";
 
 import {IERC20} from "./interface/IERC20.sol";
 
-contract PodCastNFT is ERC721URIStorage, ERC721Enumerable, Ownable {
+contract PodCastNFT is ERC721, ERC721Enumerable, Ownable {
     using Counters for Counters.Counter;
     Counters.Counter private _tokenIds;
     IERC20 public henkakuToken;
+
     uint256 public price;
     address public fundAddress;
     bytes32 private keyword;
@@ -29,8 +29,11 @@ contract PodCastNFT is ERC721URIStorage, ERC721Enumerable, Ownable {
     }
 
     PodcastKeyword private weeklyKeyword;
+
+    mapping(uint256 => string) private _tokenURIs;
     mapping(address => string[]) private roles;
     mapping(address => Attributes) private userAttribute;
+
     event BoughtMemberShipNFT(address _owner, uint256 _amount);
     event CheckedAnswer(address _by, uint256 _at);
 
@@ -42,28 +45,14 @@ contract PodCastNFT is ERC721URIStorage, ERC721Enumerable, Ownable {
         setFundAddress(_fundAddress);
     }
 
+    // override
+
     function _beforeTokenTransfer(
         address from,
         address to,
         uint256 tokenId
     ) internal override(ERC721, ERC721Enumerable) {
         super._beforeTokenTransfer(from, to, tokenId);
-    }
-
-    function _burn(uint256 tokenId)
-        internal
-        override(ERC721, ERC721URIStorage)
-    {
-        super._burn(tokenId);
-    }
-
-    function tokenURI(uint256 tokenId)
-        public
-        view
-        override(ERC721, ERC721URIStorage)
-        returns (string memory)
-    {
-        return super.tokenURI(tokenId);
     }
 
     function supportsInterface(bytes4 interfaceId)
@@ -75,10 +64,18 @@ contract PodCastNFT is ERC721URIStorage, ERC721Enumerable, Ownable {
         return super.supportsInterface(interfaceId);
     }
 
-    function setPrice(uint256 _price) public onlyOwner {
-        require(_price >= 1e18, "price must be higher than 1e18 wei");
-        price = _price;
+    function tokenURI(uint256 tokenId)
+        public
+        view
+        virtual
+        override
+        hasTokenId(tokenId)
+        returns (string memory)
+    {
+        return _tokenURIs[tokenId];
     }
+
+    // modifier
 
     modifier onlyNoneHolder(address _address) {
         require(balanceOf(_address) == 0, "Address Holds NFT");
@@ -90,6 +87,14 @@ contract PodCastNFT is ERC721URIStorage, ERC721Enumerable, Ownable {
         _;
     }
 
+    modifier hasTokenId(uint256 _tokenId) {
+        require(
+            _exists(_tokenId),
+            "ERC721Metadata: URI set of nonexistent token"
+        );
+        _;
+    }
+
     modifier onlyValidData(string memory _imageURI, string memory _name) {
         // TODO check _imageURL is really valid or not. we might check file extension for example
         require(
@@ -98,6 +103,21 @@ contract PodCastNFT is ERC721URIStorage, ERC721Enumerable, Ownable {
         );
         require(bytes(_name).length > 0, "Invalid name: name cannot be blank");
         _;
+    }
+
+    // function
+
+    function _setTokenURI(uint256 tokenId, string memory _tokenURI)
+        internal
+        virtual
+        hasTokenId(tokenId)
+    {
+        _tokenURIs[tokenId] = _tokenURI;
+    }
+
+    function setPrice(uint256 _price) public onlyOwner {
+        require(_price >= 1e18, "price must be higher than 1e18 wei");
+        price = _price;
     }
 
     function getRoles(address _address) public view returns (string[] memory) {

--- a/contracts/PodCastNFT.sol
+++ b/contracts/PodCastNFT.sol
@@ -105,7 +105,7 @@ contract PodCastNFT is ERC721, ERC721Enumerable, Ownable {
         _;
     }
 
-    // function
+    // internal function
 
     function _setTokenURI(uint256 tokenId, string memory _tokenURI)
         internal
@@ -113,63 +113,6 @@ contract PodCastNFT is ERC721, ERC721Enumerable, Ownable {
         hasTokenId(tokenId)
     {
         _tokenURIs[tokenId] = _tokenURI;
-    }
-
-    function setPrice(uint256 _price) public onlyOwner {
-        require(_price >= 1e18, "price must be higher than 1e18 wei");
-        price = _price;
-    }
-
-    function getRoles(address _address) public view returns (string[] memory) {
-        return roles[_address];
-    }
-
-    function hasRoleOf(address _address, string memory _role)
-        public
-        view
-        returns (bool)
-    {
-        string[] memory _roles = roles[_address];
-        for (uint256 i = 0; i < _roles.length; i++) {
-            if (keccak256(bytes(_roles[i])) == keccak256(bytes(_role))) {
-                return true;
-            }
-        }
-        return false;
-    }
-
-    function setRoles(address _to, string[] memory _roles)
-        public
-        onlyOwner
-        onlyHolder(_to)
-    {
-        roles[_to] = _roles;
-    }
-
-    function addRole(address _to, string memory _role)
-        public
-        onlyOwner
-        onlyHolder(_to)
-    {
-        roles[_to].push(_role);
-    }
-
-    // TODO implement updateNFT func which holder can change their name, imageURL by them self
-    function updateOwnNFT(string memory _imageURI, string memory name) public {}
-
-    function updateNFT(
-        uint256 tokenId,
-        string memory _imageURI,
-        string[] memory _roles,
-        string memory _point
-    ) public onlyOwner {
-        string memory finalTokenUri = getTokenURI(
-            tokenId,
-            _imageURI,
-            _roles,
-            _point
-        );
-        _setTokenURI(tokenId, finalTokenUri);
     }
 
     function getTokenURI(
@@ -180,16 +123,6 @@ contract PodCastNFT is ERC721, ERC721Enumerable, Ownable {
     ) internal view returns (string memory) {
         // TODO
         return "";
-    }
-
-    function mint(
-        string memory _imageURI,
-        string[] memory _roles,
-        string memory _point,
-        address _to
-    ) public onlyOwner returns (uint256) {
-        require(balanceOf(_to) == 0, "User has had already a membership NFT");
-        return _mint(_imageURI, _roles, _point, _to);
     }
 
     function _mint(
@@ -213,23 +146,52 @@ contract PodCastNFT is ERC721, ERC721Enumerable, Ownable {
         return newItemId;
     }
 
-    function mintWithHenkaku(
+    // admin function
+
+    function setPrice(uint256 _price) public onlyOwner {
+        require(_price >= 1e18, "price must be higher than 1e18 wei");
+        price = _price;
+    }
+
+    function setRoles(address _to, string[] memory _roles)
+        public
+        onlyOwner
+        onlyHolder(_to)
+    {
+        roles[_to] = _roles;
+    }
+
+    function addRole(address _to, string memory _role)
+        public
+        onlyOwner
+        onlyHolder(_to)
+    {
+        roles[_to].push(_role);
+    }
+
+    function updateNFT(
+        uint256 tokenId,
         string memory _imageURI,
-        string memory _name,
-        uint256 _amount
-    ) public onlyValidData(_imageURI, _name) onlyNoneHolder(msg.sender) {
-        require(_amount >= price, "Not Enough Henkaku");
-        bool success = henkakuToken.transferFrom(
-            msg.sender,
-            address(this),
-            _amount
+        string[] memory _roles,
+        string memory _point
+    ) public onlyOwner {
+        string memory finalTokenUri = getTokenURI(
+            tokenId,
+            _imageURI,
+            _roles,
+            _point
         );
-        require(success, "Token transfer failed");
-        string[] memory _roles = new string[](2);
-        _roles[0] = "MEMBER";
-        _roles[1] = "MINTER";
-        _mint(_imageURI, _roles, "1000", msg.sender); // FIXME _point should be numeric type
-        emit BoughtMemberShipNFT(msg.sender, _amount);
+        _setTokenURI(tokenId, finalTokenUri);
+    }
+
+    function mint(
+        string memory _imageURI,
+        string[] memory _roles,
+        string memory _point,
+        address _to
+    ) public onlyOwner returns (uint256) {
+        require(balanceOf(_to) == 0, "User has had already a membership NFT");
+        return _mint(_imageURI, _roles, _point, _to);
     }
 
     function setFundAddress(address _fundAddress) public onlyOwner {
@@ -254,6 +216,48 @@ contract PodCastNFT is ERC721, ERC721Enumerable, Ownable {
             startedAt,
             keccak256(abi.encodePacked(_keyword))
         );
+    }
+
+    // public function
+
+    function getRoles(address _address) public view returns (string[] memory) {
+        return roles[_address];
+    }
+
+    function hasRoleOf(address _address, string memory _role)
+        public
+        view
+        returns (bool)
+    {
+        string[] memory _roles = roles[_address];
+        for (uint256 i = 0; i < _roles.length; i++) {
+            if (keccak256(bytes(_roles[i])) == keccak256(bytes(_role))) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    // TODO implement updateNFT func which holder can change their name, imageURL by them self
+    function updateOwnNFT(string memory _imageURI, string memory name) public {}
+
+    function mintWithHenkaku(
+        string memory _imageURI,
+        string memory _name,
+        uint256 _amount
+    ) public onlyValidData(_imageURI, _name) onlyNoneHolder(msg.sender) {
+        require(_amount >= price, "Not Enough Henkaku");
+        bool success = henkakuToken.transferFrom(
+            msg.sender,
+            address(this),
+            _amount
+        );
+        require(success, "Token transfer failed");
+        string[] memory _roles = new string[](2);
+        _roles[0] = "MEMBER";
+        _roles[1] = "MINTER";
+        _mint(_imageURI, _roles, "1000", msg.sender); // FIXME _point should be numeric type
+        emit BoughtMemberShipNFT(msg.sender, _amount);
     }
 
     function getUserAttributes(address _of)

--- a/package.json
+++ b/package.json
@@ -4,7 +4,7 @@
     "deploy-local": "npx hardhat run scripts/deploy.js --network localhost",
     "deploy-rinkeby": "npx hardhat run scripts/deploy.js --network rinkeby",
     "verify-rinkeby": "npx hardhat verify --constructor-args arguments.js --network rinkeby",
-    "test": "yarn run hardhat test",
+    "test": "npx hardhat test",
     "lint:sol": "yarn solhint --fix --max-warnings 5 \"contracts/**/*.sol\"",
     "lint:js": "eslint --fix './{test,script}/**/*.{ts,js}' && prettier --write './{test,script}/**/*.{ts,js}'",
     "lint": "yarn lint:sol && yarn lint:js"

--- a/test/mint.js
+++ b/test/mint.js
@@ -236,6 +236,7 @@ describe('PodCastNFT', function () {
       expect(await henkakuToken.balanceOf(contract.address)).to.be.eq(
         ethers.utils.parseUnits('1000', 18)
       )
+      expect(await contract.totalSupply()).to.eq(1)
     })
 
     it('mint With henkaku token by alice', async () => {
@@ -262,6 +263,7 @@ describe('PodCastNFT', function () {
       expect(await henkakuToken.balanceOf(contract.address)).to.be.eq(
         ethers.utils.parseUnits('1000', 18)
       )
+      expect(await contract.totalSupply()).to.eq(1)
     })
   })
   describe('setFundAddress', () => {

